api/v1/publication_types.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright The CloudNativePG Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PublicationReclaimPolicy describes a policy for end-of-life maintenance of Publications.
+// +enum
+type PublicationReclaimPolicy string
+
+const (
+	// PublicationReclaimDelete means the publication will be deleted from Kubernetes on release
+	// from its claim.
+	PublicationReclaimDelete PublicationReclaimPolicy = "delete"
+
+	// PublicationReclaimRetain means the publication will be left in its current phase for manual
+	// reclamation by the administrator. The default policy is Retain.
+	PublicationReclaimRetain PublicationReclaimPolicy = "retain"
+)
+
+// PublicationSpec defines the desired state of Publication
+type PublicationSpec struct {
+	// The corresponding cluster
+	ClusterRef corev1.LocalObjectReference `json:"cluster"`
+
+	// The name inside PostgreSQL
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="name is immutable"
+	Name string `json:"name"`
+
+	// The name of the database
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="dbname is immutable"
+	DBName string `json:"dbname"`
+
+	// The owner
+	// +optional
+	Owner string `json:"owner,omitempty"`
+
+	// Parameters
+	// +optional
+	Parameters map[string]string `json:"parameters,omitempty"`
+
+	// Publication target
+	Target PublicationTarget `json:"target"`
+
+	// The policy for end-of-life maintenance of this publication
+	// +kubebuilder:validation:Enum=delete;retain
+	// +kubebuilder:default:=retain
+	// +optional
+	ReclaimPolicy PublicationReclaimPolicy `json:"publicationReclaimPolicy,omitempty"`
+}
+
+// PublicationTarget is what this publication should publish
+// +kubebuilder:validation:XValidation:rule="(has(self.allTables) && !has(self.objects)) || (!has(self.allTables) && has(self.objects))",message="allTables and objects are mutually exclusive"
+type PublicationTarget struct {
+	// All tables should be published
+	// +optional
+	AllTables bool `json:"allTables,omitempty"`
+
+	// Just the following schema objects
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="!(self.exists(o, has(o.table) && has(o.table.columns)) && self.exists(o, has(o.tablesInSchema)))",message="specifying a column list when the publication also publishes tablesInSchema is not supported"
+	// +kubebuilder:validation:MaxItems=100000
```

Comment:
This may look weird, but it is required otherwise the CRD cannot be loaded in the API server. The error would be
```
The CustomResourceDefinition "publications.postgresql.cnpg.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[target].properties[objects].x-kubernetes-validations[0].rule: Forbidden: estimated rule cost exceeds budget by factor of 1.153433x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
```